pkg/master/reconcilers/lease.go

```diff
@@ -205,9 +207,9 @@ func (r *leaseEndpointReconciler) doReconcile(serviceName string, endpointPorts
 
 	if !formatCorrect || !ipCorrect {
 		// repopulate the addresses according to the expected IPs from etcd
-		e.Subsets[0].Addresses = make([]corev1.EndpointAddress, len(masterIPs))
-		for ind, ip := range masterIPs {
-			e.Subsets[0].Addresses[ind] = corev1.EndpointAddress{IP: ip}
```

Comment:
this code will still cause <= 1.15 and >= 1.16 API servers to duel over the content of the endpoints. 1.15 servers will only copy the IP address from the leases, while 1.16 servers will copy all content. That means the kubernetes service endpoints will flip back and forth between having a node name and not having one, and will not settle.

In doReconcile, I think we have to check if *all* the API server leases contain a NodeName. If they do, that means all API servers are nodeName aware, and we can consider the nodeNames in checkEndpointSubsetFormatWithLease and populate them in `e.Subsets[0].Addresses[ind] = corev1.EndpointAddress{IP: epa.IP, NodeName: epa.NodeName}`

If not all API server leases contain a NodeName, we could be serving alongside <= 1.15 servers, so we should *ignore* node name in checkEndpointSubsetFormatWithLease and continue setting `e.Subsets[0].Addresses[ind] = corev1.EndpointAddress{IP: epa.IP}`

This needs to be well unit tested to make sure we handle those cases correctly.